File: python_backend/routes/goals.py
"""
routes/goals.py
Роуты для работы с целями знакомства
"""
from fastapi import APIRouter, Query, HTTPException, Body
from typing import List, Dict, Optional
from pydantic import BaseModel
from db_utils import db_get, db_run, safe_json_parse
import json

router = APIRouter()


class UpdateGoalsRequest(BaseModel):
    userId: str
    goals: Optional[List[str]] = None
    goal: Optional[str] = None


@router.get("/goals")
async def get_goals(userId: str = Query(..., description="ID пользователя")):
    """Получить массив целей пользователя"""
    if not userId:
        raise HTTPException(status_code=400, detail="userId обязателен")
    
    try:
        row = await db_get("SELECT goals FROM users WHERE userId = ?", [userId])
        goals_arr = safe_json_parse(row.get("goals") if row else None, [])
        return {"success": True, "goals": goals_arr}
    except Exception as e:
        print(f"GET /api/goals error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/goals")
async def update_goals(data: Dict = Body(...)):
    """Заменить весь массив целей"""
    userId = data.get("userId")
    goals = data.get("goals")
    
    if not userId:
        raise HTTPException(status_code=400, detail="userId обязателен")
    if not isinstance(goals, list):
        raise HTTPException(status_code=400, detail="goals должен быть массивом")
    
    try:
        goals_str = json.dumps(goals)
        await db_run("UPDATE users SET goals = ? WHERE userId = ?", [goals_str, userId])
        return {"success": True, "goals": goals}
    except Exception as e:
        print(f"POST /api/goals error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/updateGoals")
async def update_goals_advanced(data: UpdateGoalsRequest):
    """Обновить цели: заменить весь массив или добавить одну цель"""
    if not data.userId:
        raise HTTPException(status_code=400, detail="userId обязателен")
    
    try:
        if data.goals is not None and isinstance(data.goals, list):
            # Заменить весь массив
            goals_str = json.dumps(data.goals)
            await db_run("UPDATE users SET goals = ? WHERE userId = ?", [goals_str, data.userId])
            return {"success": True, "goals": data.goals}
        elif data.goal:
            # Добавить одну цель
            row = await db_get("SELECT goals FROM users WHERE userId = ?", [data.userId])
            goals_arr = safe_json_parse(row.get("goals") if row else None, [])
            if data.goal not in goals_arr:
                goals_arr.append(data.goal)
            goals_str = json.dumps(goals_arr)
            await db_run("UPDATE users SET goals = ? WHERE userId = ?", [goals_str, data.userId])
            return {"success": True, "goals": goals_arr}
        else:
            raise HTTPException(status_code=400, detail="Параметр goals должен быть массивом или goal строкой")
    except HTTPException:
        raise
    except Exception as e:
        print(f"POST /api/updateGoals error: {e}")
        raise HTTPException(status_code=500, detail=str(e))

File: _archive/nodejs_backend/routes/goals.js
const express = require('express');

function goalsRouter(db) {
  const router = express.Router();

  // Парсим JSON тело
  router.use(express.json());

  // GET /api/goals?userId=... — получить массив целей
  router.get('/goals', (req, res) => {
    const { userId } = req.query;
    if (!userId) {
      console.warn('GET /api/goals missing userId');
      return res.status(400).json({ success: false, error: 'userId обязателен' });
    }
    db.get(
      'SELECT goals FROM users WHERE userId = ?',
      [userId],
      (err, row) => {
        if (err) {
          console.error(`GET /api/goals DB error: ${err.message}`);
          return res.status(500).json({ success: false, error: err.message });
        }
        let goalsArr = [];
        try {
          goalsArr = JSON.parse(row?.goals || '[]');
        } catch (parseErr) {
          console.error(`GET /api/goals parse error: ${parseErr.message}`);
        }
        res.json({ success: true, goals: goalsArr });
      }
    );
  });

  // POST /api/goals — заменить весь массив целей (для совместимости с фронтендом)
  router.post('/goals', (req, res) => {
    const { userId, goals } = req.body;
    if (!userId) {
      console.warn('POST /api/goals missing userId');
      return res.status(400).json({ success: false, error: 'userId обязателен' });
    }
    if (!Array.isArray(goals)) {
      return res.status(400).json({ success: false, error: 'goals должен быть массивом' });
    }
    const goalsStr = JSON.stringify(goals);
    db.run(
      'UPDATE users SET goals = ? WHERE userId = ?',
      [goalsStr, userId],
      function(err) {
        if (err) {
          console.error(`POST /api/goals update error: ${err.message}`);
          return res.status(500).json({ success: false, error: err.message });
        }
        console.log(`✅ Цели обновлены для userId: ${userId}, affected rows: ${this.changes}`);
        res.json({ success: true, goals });
      }
    );
  });

  // POST /api/updateGoals — заменить весь массив или добавить одну цель
  router.post('/updateGoals', (req, res) => {
    const { userId, goals, goal } = req.body;
    if (!userId) {
      console.warn('POST /api/goals missing userId');
      return res.status(400).json({ success: false, error: 'userId обязателен' });
    }
    // если прислали массив goals
    if (Array.isArray(goals)) {
      const goalsStr = JSON.stringify(goals);
      db.run(
        'UPDATE users SET goals = ? WHERE userId = ?',
        [goalsStr, userId],
        function(err) {
          if (err) {
            console.error(`POST /api/goals update error: ${err.message}`);
            return res.status(500).json({ success: false, error: err.message });
          }
          res.json({ success: true, goals });
        }
      );
    } else if (typeof goal === 'string') {
      // добавить одну цель
      db.get(
        'SELECT goals FROM users WHERE userId = ?',
        [userId],
        (err, row) => {
          if (err) {
            console.error(`POST /api/goals select error: ${err.message}`);
            return res.status(500).json({ success: false, error: err.message });
          }
          let arr = [];
          try {
            arr = JSON.parse(row?.goals || '[]');
          } catch {
            arr = [];
          }
          arr.push(goal);
          const goalsStr2 = JSON.stringify(arr);
          db.run(
            'UPDATE users SET goals = ? WHERE userId = ?',
            [goalsStr2, userId],
            function(err2) {
              if (err2) {
                console.error(`POST /api/goals insert error: ${err2.message}`);
                return res.status(500).json({ success: false, error: err2.message });
              }
              res.json({ success: true, goals: arr });
            }
          );
        }
      );
    } else {
      return res
        .status(400)
        .json({ success: false, error: 'Параметр goals должен быть массивом или goal строкой' });
    }
  });

  return router;
}

module.exports = goalsRouter;